tests/unit/fem/test_assembly_levels.cpp

```diff
@@ -292,4 +292,234 @@ TEST_CASE("L2 Assembly Levels", "[AssemblyLevel], [PartialAssembly], [CUDA]")
+   {
+      for (int jj=I1[i]; jj<I1[i+1]; ++jj)
+      {
+         int j = J1[jj];
+         if (cmap1)
+         {
+            if (cmap2inv->count(cmap1[j]) > 0)
+            {
+               j = (*cmap2inv)[cmap1[j]];
+            }
+            else
+            {
+               error = std::max(error, std::fabs(V1[jj]));
+               continue;
+            }
+         }
+         error = std::max(error, std::fabs(V1[jj] - A2(i,j)));
+      }
+   }
+
+   REQUIRE(error == MFEM_Approx(0.0, 1e-10));
+}
+
+#ifdef MFEM_USE_MPI
+
+void TestSameMatrices(HypreParMatrix &A1, const HypreParMatrix &A2)
+{
+   HYPRE_BigInt *cmap1, *cmap2;
+   SparseMatrix diag1, offd1, diag2, offd2;
+
+   A1.GetDiag(diag1);
+   A2.GetDiag(diag2);
+   A1.GetOffd(offd1, cmap1);
+   A2.GetOffd(offd2, cmap2);
+
+   TestSameMatrices(diag1, diag2);
+
+   if (cmap1)
+   {
+      std::unordered_map<HYPRE_BigInt,int> cmap2inv;
+      for (int i=0; i<offd2.Width(); ++i) { cmap2inv[cmap2[i]] = i; }
+      TestSameMatrices(offd1, offd2, cmap1, &cmap2inv);
+   }
+   else
+   {
+      TestSameMatrices(offd1, offd2);
+   }
+}
+
+void TestSameHypreMatrices(OperatorHandle &A1, OperatorHandle &A2)
+{
+   HypreParMatrix *M1 = A1.Is<HypreParMatrix>();
+   HypreParMatrix *M2 = A2.Is<HypreParMatrix>();
+
+   REQUIRE(M1 != NULL);
+   REQUIRE(M2 != NULL);
+
+   TestSameMatrices(*M1, *M2);
+   TestSameMatrices(*M2, *M1);
+}
+
+void TestSameSparseMatrices(OperatorHandle &A1, OperatorHandle &A2)
+{
+   SparseMatrix *M1 = A1.Is<SparseMatrix>();
+   SparseMatrix *M2 = A2.Is<SparseMatrix>();
+
+   REQUIRE(M1 != NULL);
+   REQUIRE(M2 != NULL);
+
+   TestSameMatrices(*M1, *M2);
+   TestSameMatrices(*M2, *M1);
+}
+
+TEST_CASE("Serial H1 Full Assembly", "[AssemblyLevel], [CUDA]")
+{
+   auto order = GENERATE(1, 2, 3);
+   auto mesh_fname = GENERATE(
+                        "../../data/star.mesh",
+                        "../../data/fichera.mesh"
+                     );
+
+   Mesh mesh(mesh_fname);
+   int dim = mesh.Dimension();
+
+   H1_FECollection fec(order, dim);
+   FiniteElementSpace fespace(&mesh, &fec);
+
+   Array<int> ess_tdof_list;
+   fespace.GetBoundaryTrueDofs(ess_tdof_list);
+
+   BilinearForm a_fa(&fespace);
+   BilinearForm a_legacy(&fespace);
+
+   a_fa.SetAssemblyLevel(AssemblyLevel::FULL);
+   a_legacy.SetAssemblyLevel(AssemblyLevel::LEGACY);
+
+   a_fa.AddDomainIntegrator(new DiffusionIntegrator);
+   a_legacy.AddDomainIntegrator(new DiffusionIntegrator);
+
+   a_fa.Assemble();
+
+   a_legacy.SetDiagonalPolicy(Operator::DIAG_ONE);
+   a_legacy.Assemble();
+   a_legacy.Finalize();
+
+   OperatorHandle A_fa, A_legacy;
+   // Test that FormSystemMatrix gives the same result
+   a_fa.FormSystemMatrix(ess_tdof_list, A_fa);
+   a_legacy.FormSystemMatrix(ess_tdof_list, A_legacy);
+
+   TestSameSparseMatrices(A_fa, A_legacy);
+
+   // Test that FormLinearSystem gives the same result
+   GridFunction x1(&fespace);
+   LinearForm b1(&fespace);
+
+   // x1.Randomize(1);
```

Comment:
Remove this?

Reply:
I used the random initialization instead of the constant one in [c81506c](https://github.com/mfem/mfem/pull/3056/commits/c81506c238ec3d5ed17bcbd7871402aa6c3053dd)